File: dotnet/server/src/SessionStore.cs
namespace SmooAI.SmoothOperator.Server;

/// <summary>A conversation session: the unit the protocol's create/get operate on.</summary>
public sealed record StoredSession(
    string SessionId,
    string ConversationId,
    string AgentId,
    string AgentName,
    string UserParticipantId,
    string AgentParticipantId,
    string? UserEmail = null);

public enum MessageDirection
{
    /// <summary>From the user.</summary>
    Inbound,

    /// <summary>From the agent.</summary>
    Outbound,
}

/// <summary>
/// One logged conversation message. <see cref="CreatedAt"/> is an init-only property (not a
/// positional parameter) on purpose: it was added for <c>get_conversation_messages</c> paging and a
/// positional param would break every downstream host that constructs a <c>StoredMessage</c>. Stores
/// that persist a real timestamp set it explicitly; the rest get "now", which is correct for an
/// in-process store that appends as it goes. th-30a8a7.
/// </summary>
public sealed record StoredMessage(string Id, string ConversationId, MessageDirection Direction, string Text)
{
    public DateTimeOffset CreatedAt { get; init; } = DateTimeOffset.UtcNow;
}

/// <summary>
/// One row of the conversation-list / resume surface: identity, last activity, message count,
/// and the first inbound (user) message text — enough for the dispatcher to build a sidebar
/// title without a second store roundtrip. The C# analog of the Rust <c>list_conversations</c>'
/// per-conversation peek and the Go <c>ConversationSummary</c>; title formatting (markdown strip,
/// truncation, ISO timestamp) is the dispatcher's job. <c>FirstInboundText</c> is <c>null</c> when
/// the conversation has no inbound message (the title falls back to a generic name). th-d5b446.
/// </summary>
public sealed record ConversationSummary(string ConversationId, DateTimeOffset UpdatedAt, int MessageCount, string? FirstInboundText);

/// <summary>
/// Which user's conversations a read may see. A store MUST honour this in its query, not after it —
/// filtering a page in memory after a LIMIT silently returns short or empty pages.
/// <para>
/// This is a type rather than a nullable <c>string</c> on purpose: "no filter" means "every user's
/// conversations", and that must be spelled out (<see cref="Unscoped"/>) instead of falling out of a
/// forgotten <c>null</c>. th-966fab.
/// </para>
/// </summary>
public readonly record struct ConversationScope
{
    private ConversationScope(string? userEmail, bool unscoped)
    {
        UserEmail = userEmail;
        IsUnscoped = unscoped;
    }

    /// <summary>The owning user's email when scoped; <c>null</c> for <see cref="Unscoped"/>/<see cref="None"/>.</summary>
    public string? UserEmail { get; }

    /// <summary>
    /// EVERY user's conversations. Legitimate ONLY on a server with no auth configured (single-tenant
    /// local/dev, where there is no notion of a user). Never reachable from an authenticated request.
    /// </summary>
    public bool IsUnscoped { get; }

    /// <summary>No conversations at all — an authenticated caller whose identity carries no email.</summary>
    public bool IsEmpty => !IsUnscoped && string.IsNullOrEmpty(UserEmail);

    public static ConversationScope Unscoped { get; } = new(null, unscoped: true);

    /// <summary>Nothing matches. The fail-closed scope. </summary>
    public static ConversationScope None { get; } = new(null, unscoped: false);

    public static ConversationScope ForUser(string userEmail) => new(userEmail, unscoped: false);
}

/// <summary>
/// Persistence for sessions + conversation message logs — the C# analog of the Rust
/// <c>StorageAdapter</c>'s session/conversation/message surface (and, like it, async). The
/// bundled <see cref="InMemorySessionStore"/> is the reference store; a Postgres adapter
/// (<c>SmooAI.SmoothOperator.Server.Postgres</c>) implements the same interface for durability.
/// </summary>
public interface ISessionStore
{
    Task<StoredSession> CreateSessionAsync(string agentId, string? userName, string? userEmail, CancellationToken cancellationToken = default);

    /// <summary>
    /// Mint a session bound to an existing conversation when <paramref name="conversationId"/> is
    /// non-empty AND known (reuses its message log so subsequent turns append to it and the runner
    /// replays its history); an empty or unknown <paramref name="conversationId"/> mints a fresh
    /// conversation — identical to <see cref="CreateSessionAsync"/>. The resume substrate behind
    /// <c>create_conversation_session</c>'s optional <c>conversationId</c>. th-d5b446.
    /// </summary>
    Task<StoredSession> ResumeSessionAsync(string agentId, string? userName, string? userEmail, string? conversationId, CancellationToken cancellationToken = default);

    /// <summary>
    /// A summary per conversation that has at least one message (empty conversations — every
    /// page-load currently mints one — are filtered out), in no particular order; the dispatcher
    /// sorts most-recent-first and caps. The C# analog of the Rust storage list-conversations +
    /// per-conversation peek and the Go <c>ListConversations</c>. th-d5b446.
    /// <para>
    /// SECURITY (th-966fab): <paramref name="scope"/> is REQUIRED and MUST be applied inside the
    /// query. <see cref="ConversationScope.ForUser"/> returns only conversations owned by that email;
    /// <see cref="ConversationScope.None"/> returns nothing; <see cref="ConversationScope.Unscoped"/>
    /// returns every user's conversations and is legitimate ONLY on a server with no auth configured.
    /// Ignoring the scope re-opens a cross-user data leak.
    /// </para>
    /// </summary>
    Task<IReadOnlyList<ConversationSummary>> ListConversationsAsync(ConversationScope scope, CancellationToken cancellationToken = default);

    /// <summary>
    /// Whether <paramref name="userEmail"/> owns <paramref name="conversationId"/> — the ownership
    /// gate behind resume and <c>get_conversation_messages</c>.
    /// <para>
    /// SECURITY (th-966fab): returns <c>false</c> for a conversation that does not exist, one owned by
    /// another user, AND one with no recorded owner (data written before per-user scoping existed).
    /// Collapsing all three into one answer is deliberate — a caller cannot tell "not yours" from
    /// "never existed", so this cannot be used to enumerate other users' conversation ids.
    /// </para>
    /// </summary>
    Task<bool> ConversationBelongsToUserAsync(string conversationId, string userEmail, CancellationToken cancellationToken = default);

    Task<StoredSession?> GetSessionAsync(string sessionId, CancellationToken cancellationToken = default);

    Task<StoredMessage> AppendMessageAsync(string conversationId, MessageDirection direction, string text, CancellationToken cancellationToken = default);

    /// <summary>The most recent <paramref name="limit"/> messages for a conversation, oldest first.</summary>
    Task<IReadOnlyList<StoredMessage>> ListMessagesAsync(string conversationId, int limit, CancellationToken cancellationToken = default);

    /// <summary>The persisted conversation-workflow step pointer for a conversation, or <c>null</c>
    /// when none has been recorded (a fresh conversation starts on the workflow's first step).
    /// Mirrors the monorepo graph state's <c>currentStepId</c>, persisted so a workflow advances
    /// across turns (and connections).</summary>
    Task<string?> GetWorkflowStepAsync(string conversationId, CancellationToken cancellationToken = default);

    /// <summary>Record the conversation's current workflow step (upsert). Called after the judge
    /// advances the pointer at the end of a turn.</summary>
    Task SetWorkflowStepAsync(string conversationId, string stepId, CancellationToken cancellationToken = default);

    /// <summary>Whether this conversation's caller is identity-verified (the persisted
    /// <c>otpVerified</c> bit — the C# analog of the Rust session's <c>metadata.otpVerified</c>).
    /// <c>false</c> for a fresh or unknown conversation. Threaded into the <c>end_user</c> auth gate
    /// via <see cref="StoreSessionAuthenticator"/> so a verified caller's gated tools run.</summary>
    Task<bool> GetSessionAuthenticatedAsync(string conversationId, CancellationToken cancellationToken = default);

    /// <summary>Mark this conversation's caller identity-verified (or clear it). Called after a
    /// successful <c>verify_otp</c>. Upsert; no-op semantics for an unknown conversation are fine
    /// (the bit simply reads back on the next turn).</summary>
    Task SetSessionAuthenticatedAsync(string conversationId, bool verified, CancellationToken cancellationToken = default);
}

/// <summary>
/// An <see cref="ISessionAuthenticator"/> backed by the session store's persisted <c>otpVerified</c>
/// bit — the default when a host wires no authenticator of its own. Fails closed for any conversation
/// that never completed <c>verify_otp</c> (reads <c>false</c>), so an unwired server is unchanged; a
/// verified session reads <c>true</c> and its <c>end_user</c> tools run. Mirrors the Rust reference
/// threading <c>session_authenticated</c> (from session metadata) into <c>build_auth_gate</c>.
/// </summary>
public sealed class StoreSessionAuthenticator : ISessionAuthenticator
{
    private readonly ISessionStore _store;

    public StoreSessionAuthenticator(ISessionStore store) => _store = store;

    public Task<bool> IsAuthenticatedAsync(string conversationId, CancellationToken cancellationToken = default) =>
        _store.GetSessionAuthenticatedAsync(conversationId, cancellationToken);
}

/// <summary>In-process <see cref="ISessionStore"/>. The C# analog of the Rust in-memory adapter.</summary>
public sealed class InMemorySessionStore : ISessionStore
{
    private readonly object _gate = new();
    private readonly Dictionary<string, StoredSession> _sessions = new();
    private readonly Dictionary<string, List<StoredMessage>> _messages = new();
    private readonly Dictionary<string, string> _workflowSteps = new();
    private readonly HashSet<string> _authenticated = new();

    // Each conversation's last activity (creation, then every append) — the sort key + updatedAt
    // field for ListConversations. th-d5b446.
    private readonly Dictionary<string, DateTimeOffset> _updatedAt = new();

    // Each conversation's owning user email, stamped when the conversation is minted (null when the
    // creator had no identity — no-auth servers, or an authenticated principal with no email claim).
    // The scoping key for ListConversations + the ownership gate. th-966fab.
    private readonly Dictionary<string, string?> _owner = new();

    public Task<StoredSession> CreateSessionAsync(string agentId, string? userName, string? userEmail, CancellationToken cancellationToken = default) =>
        ResumeSessionAsync(agentId, userName, userEmail, null, cancellationToken);

    public Task<StoredSession> ResumeSessionAsync(string agentId, string? userName, string? userEmail, string? conversationId, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            // Resume when the caller names a known conversation (reuse its id + message log);
            // absent/unknown → a fresh conversation (byte-for-byte the old CreateSession behavior).
            var resume = !string.IsNullOrEmpty(conversationId) && _messages.ContainsKey(conversationId);
            var convId = resume ? conversationId! : Guid.NewGuid().ToString();

            var session = new StoredSession(
                SessionId: Guid.NewGuid().ToString(),
                ConversationId: convId,
                AgentId: string.IsNullOrEmpty(agentId) ? Guid.NewGuid().ToString() : agentId,
                AgentName: "smooth-agent",
                UserParticipantId: Guid.NewGuid().ToString(),
                AgentParticipantId: Guid.NewGuid().ToString(),
                UserEmail: string.IsNullOrEmpty(userEmail) ? null : userEmail);

            _sessions[session.SessionId] = session;
            // Only mint an empty log + creation timestamp on a fresh conversation — a resume keeps
            // its history and its last-activity time (bumped by the next append, not by re-binding).
            if (!resume)
            {
                _messages[convId] = new List<StoredMessage>();
                _updatedAt[convId] = DateTimeOffset.UtcNow;
                _owner[convId] = session.UserEmail; // ownership is stamped once, at mint. th-966fab.
            }
            return Task.FromResult(session);
        }
    }

    public Task<IReadOnlyList<ConversationSummary>> ListConversationsAsync(ConversationScope scope, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            var summaries = new List<ConversationSummary>();
            if (scope.IsEmpty)
            {
                // An authenticated caller with no identity email owns nothing. th-966fab.
                return Task.FromResult<IReadOnlyList<ConversationSummary>>(summaries);
            }

            foreach (var (convId, list) in _messages)
            {
                if (list.Count == 0)
                {
                    continue; // drop the empty conversations every page-load mints.
                }
                if (!scope.IsUnscoped && !OwnedBy(convId, scope.UserEmail!))
                {
                    continue; // not this user's conversation. th-966fab.
                }
                var firstInbound = list.FirstOrDefault(m => m.Direction == MessageDirection.Inbound)?.Text;
                var updatedAt = _updatedAt.TryGetValue(convId, out var t) ? t : DateTimeOffset.UtcNow;
                summaries.Add(new ConversationSummary(convId, updatedAt, list.Count, firstInbound));
            }
            IReadOnlyList<ConversationSummary> result = summaries;
            return Task.FromResult(result);
        }
    }

    public Task<bool> ConversationBelongsToUserAsync(string conversationId, string userEmail, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            return Task.FromResult(OwnedBy(conversationId, userEmail));
        }
    }

    // Unknown conversation, another user's, and one with no recorded owner are all `false` — the
    // caller cannot distinguish them, so this is not an existence oracle. Caller holds _gate.
    private bool OwnedBy(string conversationId, string userEmail) =>
        _owner.TryGetValue(conversationId, out var owner) && owner is not null && string.Equals(owner, userEmail, StringComparison.OrdinalIgnoreCase);

    public Task<StoredSession?> GetSessionAsync(string sessionId, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            return Task.FromResult(_sessions.TryGetValue(sessionId, out var session) ? session : null);
        }
    }

    public Task<StoredMessage> AppendMessageAsync(string conversationId, MessageDirection direction, string text, CancellationToken cancellationToken = default)
    {
        var message = new StoredMessage(Guid.NewGuid().ToString(), conversationId, direction, text);
        lock (_gate)
        {
            if (!_messages.TryGetValue(conversationId, out var list))
            {
                list = new List<StoredMessage>();
                _messages[conversationId] = list;
            }
            list.Add(message);
            _updatedAt[conversationId] = DateTimeOffset.UtcNow; // last activity → ListConversations sort key.
        }
        return Task.FromResult(message);
    }

    public Task<IReadOnlyList<StoredMessage>> ListMessagesAsync(string conversationId, int limit, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            IReadOnlyList<StoredMessage> result = _messages.TryGetValue(conversationId, out var list)
                ? list.TakeLast(limit).ToList()
                : Array.Empty<StoredMessage>();
            return Task.FromResult(result);
        }
    }

    public Task<string?> GetWorkflowStepAsync(string conversationId, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            return Task.FromResult(_workflowSteps.TryGetValue(conversationId, out var step) ? step : null);
        }
    }

    public Task SetWorkflowStepAsync(string conversationId, string stepId, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            _workflowSteps[conversationId] = stepId;
        }
        return Task.CompletedTask;
    }

    public Task<bool> GetSessionAuthenticatedAsync(string conversationId, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            return Task.FromResult(_authenticated.Contains(conversationId));
        }
    }

    public Task SetSessionAuthenticatedAsync(string conversationId, bool verified, CancellationToken cancellationToken = default)
    {
        lock (_gate)
        {
            if (verified)
            {
                _authenticated.Add(conversationId);
            }
            else
            {
                _authenticated.Remove(conversationId);
            }
        }
        return Task.CompletedTask;
    }
}

File: python/server/src/smooth_operator_server/session_store.py
"""Session + conversation-message persistence.

The Python analog of the C# ``ISessionStore`` / ``InMemorySessionStore`` and the
Rust storage adapter's session/message surface. The protocol's
``create_conversation_session`` / ``get_session`` operate on sessions; a turn
appends to (and replays) the conversation message log.

The bundled :class:`InMemorySessionStore` is the reference store (lost on restart);
the abstract :class:`SessionStore` is the seam a durable (Postgres/Dynamo) adapter
would implement.
"""

from __future__ import annotations

import uuid
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from threading import Lock

from .auth import normalize_email


@dataclass(frozen=True)
class StoredSession:
    """A conversation session — the unit create/get operate on."""

    session_id: str
    conversation_id: str
    agent_id: str
    agent_name: str
    user_participant_id: str
    agent_participant_id: str
    #: The caller's email captured at create time, used as the OTP delivery contact
    #: for the ``end_user`` identity flow (the Python analog of the Rust session's
    #: ``metadata.contactEmail``). ``None`` when no email was supplied — the server
    #: then can't offer OTP for this session.
    contact_email: str | None = None
    #: The AUTHENTICATED principal's email that owns this session's conversation — the
    #: ACL key (th-8fe998). Set from the connection's principal, NEVER from a client
    #: frame field. ``None`` means "no owner" — an anonymous/emailless principal, or a
    #: session predating ownership — and stays reachable by everyone (th-909995); with
    #: auth disabled (the single-tenant local flavor) ownership isn't consulted at all.
    owner_email: str | None = None


class MessageDirection(Enum):
    """Who a message came from."""

    INBOUND = "inbound"  # from the user
    OUTBOUND = "outbound"  # from the agent


@dataclass(frozen=True)
class StoredMessage:
    """One logged conversation message."""

    id: str
    conversation_id: str
    direction: MessageDirection
    text: str
    #: When the message was appended (UTC) — the ``createdAt`` field of the
    #: ``get_conversation_messages`` contract and its ``before`` paging key. Defaulted so
    #: downstream implementers of :class:`SessionStore` that build a ``StoredMessage``
    #: positionally keep working. th-89b698.
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))


@dataclass(frozen=True)
class ConversationSummary:
    """One conversation's roll-up for the ``list_conversations`` action — enough to
    render a resumable-thread list without pulling every message. The dispatcher turns
    ``first_inbound_text`` / ``updated_at`` / ``message_count`` into the wire
    ``{conversationId, title, updatedAt, messageCount}``. The Python analog of the Go
    ``ConversationSummary`` and the TS ``ConversationSummary``. th-d5b446."""

    conversation_id: str
    #: Last-activity timestamp (create, then every append) — the sort key + ``updatedAt``.
    updated_at: datetime
    #: Total messages in the conversation. The dispatcher drops empties (``0``).
    message_count: int
    #: Text of the FIRST inbound (user) message — the dispatcher's title source.
    #: ``None`` when the conversation has no inbound message (title falls back to a name).
    first_inbound_text: str | None = None


#: The reference agent's display name (mirrors the Rust ``AGENT_NAME`` and the C#
#: ``InMemorySessionStore`` default).
AGENT_NAME = "smooth-agent"


class SessionStore(ABC):
    """Persistence for sessions + conversation message logs (async, like the Rust
    adapter and the C# ``ISessionStore``)."""

    @abstractmethod
    async def create_session(
        self,
        agent_id: str,
        user_name: str | None,
        user_email: str | None,
        conversation_id: str | None = None,
        *,
        owner_email: str | None = None,
        enforced: bool = False,
    ) -> StoredSession:
        """Mint a session. When ``conversation_id`` names an EXISTING conversation that
        ``owner_email`` may reach, the new session binds to it (resume: reuses the id +
        its persisted message log, so subsequent turns append and history replays). An
        absent, unknown, **or not-reachable** id mints a fresh conversation.

        ``owner_email`` is the authenticated principal's email and must come from the
        connection's auth context, never from a client-supplied frame field. ``None``
        means either "auth disabled" or "authenticated but emailless"; ``enforced``
        distinguishes them — ``False`` (auth disabled, single tenant) skips the
        ownership check entirely, ``True`` restricts the resume to conversations owned
        by ``owner_email`` **or by nobody**. Ownerless conversations stay reachable so
        anonymous/emailless principals can resume what they themselves created;
        refusing those locked them out of their own sessions. th-909995.

        Resuming a conversation owned by ANOTHER user must be indistinguishable from
        resuming an id that never existed: both mint a fresh conversation. Do not add a
        distinct error for the not-owned case — that builds an existence oracle.
        th-d5b446, th-8fe998."""
        ...

    @abstractmethod
    async def get_session(self, session_id: str) -> StoredSession | None: ...

    @abstractmethod
    async def list_conversations(self, user_email: str | None, *, enforced: bool = False) -> list[ConversationSummary]:
        """A summary per conversation reachable by ``user_email`` that has at least one
        message (empty conversations — every page-load currently mints one — are
        dropped), in no particular order; the dispatcher sorts most-recent-first and
        caps. The Python analog of the Rust ``list_conversations_by_org`` +
        per-conversation peek. th-d5b446, th-8fe998.

        ``user_email`` is **required** — deliberately not defaulted. A default would be
        fail-OPEN and would let an implementer of this protocol ship an unscoped (i.e.
        cross-user-leaking) store without ever confronting the question.

        ``enforced=False`` is the single-tenant, auth-disabled flavor: unscoped, every
        conversation. ``enforced=True`` returns those owned by ``user_email`` plus the
        ownerless ones (anonymous/emailless/legacy), never another owner's. th-909995.

        Apply the filter in the SELECTION itself, never after the dispatcher's limit —
        filtering a limited page silently returns short or empty pages."""
        ...

    @abstractmethod
    async def append_message(self, conversation_id: str, direction: MessageDirection, text: str) -> StoredMessage: ...

    @abstractmethod
    async def list_messages(self, conversation_id: str, limit: int) -> list[StoredMessage]:
        """The most recent ``limit`` messages for a conversation, oldest first."""
        ...

    @abstractmethod
    async def get_current_step_id(self, conversation_id: str) -> str | None:
        """The conversation's current workflow-step pointer (``None`` = fresh start)."""
        ...

    @abstractmethod
    async def set_current_step_id(self, conversation_id: str, step_id: str | None) -> None:
        """Persist the conversation's workflow-step pointer (the analog of the TS
        ``state.currentStepId`` carried across turns)."""
        ...

    @abstractmethod
    async def is_session_authenticated(self, session_id: str) -> bool:
        """Whether the caller has completed OTP identity verification for this session
        (the Python analog of the Rust session's ``metadata.otpVerified``). ``False``
        for an unknown or unverified session. Threaded into the ``end_user`` auth gate
        so a verified session's gated tools run."""
        ...

    @abstractmethod
    async def set_session_authenticated(self, session_id: str, verified: bool) -> None:
        """Mark this session identity-verified (or clear it). Called after a
        successful ``verify_otp``. A no-op for an unknown session."""
        ...


class InMemorySessionStore(SessionStore):
    """In-process :class:`SessionStore` — the reference store (the C# analog of
    ``InMemorySessionStore`` / the Rust in-memory adapter). A lock guards the dicts
    so concurrent connections never corrupt them."""

    def __init__(self) -> None:
        self._gate = Lock()
        self._sessions: dict[str, StoredSession] = {}
        self._messages: dict[str, list[StoredMessage]] = {}
        #: Per-conversation last-activity time (create, then every append) — the sort key
        #: + updated_at source for list_conversations. th-d5b446.
        self._updated_at: dict[str, datetime] = {}
        #: Per-conversation owner email (the ACL key, normalized). Absent/None = owned
        #: by nobody → reachable by every principal (th-909995: anonymous and emailless
        #: principals mint exactly these, and must not be locked out of them).
        self._owners: dict[str, str | None] = {}
        #: Per-conversation workflow-step pointer (absent = fresh start / no workflow).
        self._current_step: dict[str, str] = {}
        #: Per-session OTP-verified bit (absent/False = unverified). Set by a
        #: successful ``verify_otp``; read by the ``end_user`` auth gate.
        self._authenticated: dict[str, bool] = {}

    async def create_session(
        self,
        agent_id: str,
        user_name: str | None,
        user_email: str | None,
        conversation_id: str | None = None,
        *,
        owner_email: str | None = None,
        enforced: bool = False,
    ) -> StoredSession:
        owner = normalize_email(owner_email)
        with self._gate:
            # Resume: bind to an existing conversation (reuse its id + persisted log) when
            # the caller passes a known conversationId THEY MAY REACH — theirs, or an
            # ownerless one. Unknown, absent, or owned by someone else → mint a fresh
            # one. Not-owned takes the exact same branch as never-existed, so the two are
            # indistinguishable on the wire — a separate refusal would let an attacker
            # probe for other users' ids. `enforced=False` = auth disabled (single
            # tenant): ownership isn't consulted. th-909995.
            resume = (
                bool(conversation_id)
                and conversation_id in self._messages
                and (not enforced or self._owners.get(conversation_id) in (None, owner))
            )
            conv_id = conversation_id if resume else str(uuid.uuid4())
            session = StoredSession(
                session_id=str(uuid.uuid4()),
                conversation_id=conv_id,
                agent_id=agent_id if agent_id else str(uuid.uuid4()),
                agent_name=AGENT_NAME,
                user_participant_id=str(uuid.uuid4()),
                agent_participant_id=str(uuid.uuid4()),
                contact_email=(user_email.strip() or None) if isinstance(user_email, str) else None,
                owner_email=owner,
            )
            self._sessions[session.session_id] = session
            # Only seed an empty log + timestamp on a fresh conversation — a resume keeps
            # its history (and prior last-activity time).
            if not resume:
                self._messages[conv_id] = []
                self._updated_at[conv_id] = datetime.now(timezone.utc)
                self._owners[conv_id] = owner
        return session

    async def get_session(self, session_id: str) -> StoredSession | None:
        with self._gate:
            return self._sessions.get(session_id)

    async def append_message(self, conversation_id: str, direction: MessageDirection, text: str) -> StoredMessage:
        message = StoredMessage(str(uuid.uuid4()), conversation_id, direction, text, datetime.now(timezone.utc))
        with self._gate:
            self._messages.setdefault(conversation_id, []).append(message)
            self._updated_at[conversation_id] = datetime.now(timezone.utc)
        return message

    async def list_conversations(self, user_email: str | None, *, enforced: bool = False) -> list[ConversationSummary]:
        scope = normalize_email(user_email)
        with self._gate:
            out: list[ConversationSummary] = []
            for conv_id, log in self._messages.items():
                if not log:  # drop empties — every page-load mints one
                    continue
                # Ownership filter lives HERE, in the selection — the dispatcher's limit
                # is applied to the already-filtered result. Another owner's rows are
                # dropped; ownerless ones survive for every principal (th-909995).
                if enforced and self._owners.get(conv_id) not in (None, scope):
                    continue
                # Messages are stored oldest-first, so the first inbound is the title source.
                first_inbound = next((m.text for m in log if m.direction is MessageDirection.INBOUND), None)
                out.append(
                    ConversationSummary(
                        conversation_id=conv_id,
                        updated_at=self._updated_at.get(conv_id, datetime.now(timezone.utc)),
                        message_count=len(log),
                        first_inbound_text=first_inbound,
                    )
                )
            return out

    async def list_messages(self, conversation_id: str, limit: int) -> list[StoredMessage]:
        with self._gate:
            log = self._messages.get(conversation_id, [])
            return list(log[-limit:]) if limit > 0 else list(log)

    async def get_current_step_id(self, conversation_id: str) -> str | None:
        with self._gate:
            return self._current_step.get(conversation_id)

    async def set_current_step_id(self, conversation_id: str, step_id: str | None) -> None:
        with self._gate:
            if step_id is None:
                self._current_step.pop(conversation_id, None)
            else:
                self._current_step[conversation_id] = step_id

    async def is_session_authenticated(self, session_id: str) -> bool:
        with self._gate:
            return self._authenticated.get(session_id, False)

    async def set_session_authenticated(self, session_id: str, verified: bool) -> None:
        with self._gate:
            # Only a tracked session can be verified — mirrors the Rust no-op for an
            # unknown session (a code can't authenticate something we don't hold).
            if session_id not in self._sessions:
                return
            if verified:
                self._authenticated[session_id] = True
            else:
                self._authenticated.pop(session_id, None)
